src/python/WMCore/WebTools/RESTModel.py

```diff
@@ -185,6 +185,14 @@ def wrapper(*input_args, **input_kwargs):
             if len(args) != 0:
                 # store the method name
                 method = methodKey
+                # for PUT/POST requests data may be send in body of the request
+                # we check if this is the case (no args/kwargs) and read body
+                if  verb in ['PUT', 'POST'] and \
+                    not input_args and not input_kwargs:
+                    from WMCore.Wrappers import JsonWrapper
+                    body = cherrypy.request.body.read()
+                    if  body:
+                        input_kwargs = JsonWrapper.loads( body )
```

Comment:
I think it is another issue with current implementation of RESTModel.
The POST/PUT requests can be done without any explicit arguments and client can
send request in a body.

The logic of RESTModel should adapt to this generic statement.

The line
https://github.com/vkuznet/WMCore/blob/9a68173916e5ef994f0453cabbfce7077f322648/src/python/WMCore/WebTools/RESTModel.py#L241

does not fit in generic REST model approach. It is custom to our
implementation. And, I think we should generalize it accordingly.
Moreover, line 242 and 243 is common source of misleader when something goes
wrong. The HTTP error is probably fine, but self.debug statements never
appears in log unless someone explicitly configure to use DEBUG level.

I would suggest to re-do implementation of _sanitise_input with respect to
generic REST model approach, i.e. client should be allowed to send
both POST/PUT requests with explicit arguments and using body.

My original change is at lowest level which allows to adjust
_sanitise_input implementation.

On  0, ticoann <notifications@github.com> wrote:
> > @@ -185,6 +185,14 @@ def wrapper(*input_args, **input_kwargs):
> >              if len(args) != 0:
> >                  # store the method name
> >                  method = methodKey
> > +                # for PUT/POST requests data may be send in body of the request
> > +                # we check if this is the case (no args/kwargs) and read body
> > +                if  verb in ['PUT', 'POST'] and \
> > +                    not input_args and not input_kwargs:
> > +                    from WMCore.Wrappers import JsonWrapper
> > +                    body = cherrypy.request.body.read()
> > +                    if  body:
> > +                        input_kwargs = JsonWrapper.loads( body )
> 
> Sorry I wasn't clear. What I meant is that following line checks the number of argument defined by method and compare them what is passed.
> 
> https://github.com/vkuznet/WMCore/blob/9a68173916e5ef994f0453cabbfce7077f322648/src/python/WMCore/WebTools/RESTModel.py#L241
> 
> For the parameters passed by body, might not defined in the function definition.
> i.e.
> https://github.com/vkuznet/WMCore/blob/9a68173916e5ef994f0453cabbfce7077f322648/src/python/WMCore/HTTPFrontEnd/RequestManager/ReqMgrRESTModel.py#L133
> 
> only take one argument but if it contains message body with other parameter (i.e. message) it will fail. 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/dmwm/WMCore/pull/6495/files#r49346498